src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,906 @@
+        return false;
+
+    return true;
+}
+
+static int
+_service_connect(sd_bus_message *reply, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct sol_connman_service *service = userdata;
+    struct sol_connman_pending *pending;
+
+    SOL_NULL_CHECK(service, -EINVAL);
+
+    pending = &service->pending;
+    pending->slot = sd_bus_slot_unref(pending->slot);
+
+    if (sol_bus_log_callback(reply, userdata, ret_error) < 0)
+        service->is_call_success = false;
+    else
+        service->is_call_success = true;
+
+    if (_ctx.monitor_callback.cb)
+        _ctx.monitor_callback.cb(_ctx.monitor_callback.data, service);
+
+    return 0;
+}
+
+SOL_API int
+sol_connman_service_connect(struct sol_connman_service *service)
+{
+    int r;
+    struct sol_connman_pending *pending;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+    SOL_NULL_CHECK(service, -EINVAL);
+    SOL_NULL_CHECK(service->path, -EINVAL);
+
+    pending = &service->pending;
+    service->is_call_success = false;
+
+    r = sd_bus_call_method_async(bus, &pending->slot, "net.connman", service->path,
+        "net.connman.Service", "Connect", _service_connect, service, NULL);
+    SOL_INT_CHECK(r, < 0, -EINVAL);
+
+    return 0;
+}
+
+static int
+_service_disconnect(sd_bus_message *reply, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct sol_connman_service *service = userdata;
+    struct sol_connman_pending *pending;
+
+    SOL_NULL_CHECK(service, -EINVAL);
+
+    pending = &service->pending;
+    pending->slot = sd_bus_slot_unref(pending->slot);
+
+    if (sol_bus_log_callback(reply, userdata, ret_error) < 0)
+        service->is_call_success = false;
+    else
+        service->is_call_success = true;
+
+    if (_ctx.monitor_callback.cb)
+        _ctx.monitor_callback.cb(_ctx.monitor_callback.data, service);
+
+    return 0;
+}
+
+SOL_API int
+sol_connman_service_disconnect(struct sol_connman_service *service)
+{
+    int r;
+    struct sol_connman_pending *pending;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+    SOL_NULL_CHECK(service, -EINVAL);
+    SOL_NULL_CHECK(service->path, -EINVAL);
+
+    pending = &service->pending;
+
+    if (pending->slot)
+        pending->slot = sd_bus_slot_unref(pending->slot);
+
+    service->is_call_success = false;
+
+    r = sd_bus_call_method_async(bus, &pending->slot, "net.connman", service->path,
+        "net.connman.Service", "Disconnect", _service_disconnect, service, NULL);
+    SOL_INT_CHECK(r, < 0, -EINVAL);
+
+    return 0;
+}
+
+int
+sol_connman_init(void)
+{
+    _ctx.connman = NULL;
+    _ctx.state_slot.slot = NULL;
+    _ctx.manager_slot.slot = NULL;
+    _ctx.service_slot.slot = NULL;
+    _ctx.connman_state = SOL_CONNMAN_STATE_UNKNOWN;
+    _ctx.properties_changed = NULL;
+    _ctx.monitor_callback.cb = NULL;
+    _ctx.monitor_callback.data = NULL;
+
+    return 0;
+}
+
+void
+sol_connman_shutdown(void)
```

Comment:
If the user calls shutdown without removing the monitor, a lot of resources will be leaked.